CRM/MembershipExtras/Hook/PostProcess/UpdateSubscription.php

```diff
@@ -1,296 +1,331 @@
-  }
-
-  /**
-   * Checks if payment instrument was updated by user.
-   *
-   * @return bool
-   *   True if the value was changed by user, false otherwise.
-   */
-  private function isUpdatedPaymentInstrument() {
-    $formValues = $this->form->exportValues();
-    $oldPaymentInstrument = CRM_Utils_Request::retrieve('old_payment_instrument_id', 'Integer', $this->form, TRUE);
-
-    return $formValues['cycle_day'] != $oldPaymentInstrument;
-  }
-
-}
+<?php
+
+use CRM_MembershipExtras_Service_InstallmentReceiveDateCalculator as InstallmentReceiveDateCalculator;
+
+/**
+ * Post-processes Recurring Conribution Update form.
+ */
+class CRM_MembershipExtras_Hook_PostProcess_UpdateSubscription {
+
+  /**
+   * Form to be processed.
+   *
+   * @var \CRM_Contribute_Form_UpdateSubscription
+   */
+  private $form;
+
+  /**
+   * Array with the data of the recurring contribution that is being updated.
+   *
+   * @var array
+   */
+  private $recurringContribution;
+
+  /**
+   * Object that calcuates installment receive date.
+   *
+   * @var CRM_MembershipExtras_Service_InstallmentReceiveDateCalculator
+   */
+  private $receiveDateCalculator;
+
+  /**
+   * Due date of next contribution of the payment plan.
+   *
+   * @var string
+   */
+  private $nextContributionDate;
+
+  public function __construct(CRM_Contribute_Form_UpdateSubscription $form) {
+    $this->form = $form;
+    $this->setRecurringContribution();
+  }
+
+  /**
+   * Loads data for recurring contribution identified by 'crid' parameter in
+   * http request.
+   */
+  private function setRecurringContribution() {
+    $recurringContributionID = CRM_Utils_Request::retrieve('crid', 'Integer', $this->form, TRUE);
+    $this->recurringContribution = civicrm_api3('ContributionRecur', 'get', [
+      'sequential' => 1,
+      'id' => $recurringContributionID,
+    ])['values'][0];
+  }
+
+  /**
+   * Post-processes CRM_Contribute_Form_UpdateSubscription to deal with added
+   * fields.
+   */
+  public function postProcess() {
+    $this->updateMemberships();
+    $this->updateRelatedInstallments();
+    $this->updateRecurringContribution();
+    $this->updateSubscriptionLineItems();
+  }
+
+  /**
+   * Updates data for recurring contribution.
+   */
+  private function updateRecurringContribution() {
+    $autoRenew = $this->form->getElementValue('auto_renew');
+    $params = [
+      'id' => $this->recurringContribution['id'],
+      'auto_renew' => $autoRenew,
+    ];
+
+    if ($this->isUpdatedCycleDate()) {
+      $params['next_sched_contribution_date'] = $this->nextContributionDate;
+      $params['start_date'] = $this->getReceiveDateOfFirstInstallment() ?: $params['start_date'];
+    }
+
+    civicrm_api3('ContributionRecur', 'create', $params);
+  }
+
+  /**
+   * Obteins the start date of the first contribution.
+   *
+   * @return string
+   */
+  private function getReceiveDateOfFirstInstallment() {
+    $contributions = $this->getContributions();
```

Comment:
you don't need to get all the contribution to filter out the first one, just alter the method to get the first one directly.

Reply:
`$this->getContributions()` is also used when updating installments. So I've brought into this method the call to API to obtain only the first contribution.